### PR TITLE
feat(user feedback): display a button in a window that presents a vc

### PR DESCRIPTION
**Pull request by armcknight**
#skip-changelog; for #4270

Deliver a complete styled button like we have in the javascript SDK docs that opens a modal view where the form will go in a subsequent PR.
image

Check the linked issue for updates on what it looks like with various config settings and device settings like for accessibility and dark mode.

### CODE REVIEW

File: Sources/Sentry/Public/SentryOptions.h [hunk 1]
```diff
@@ -164,7 +164,7 @@ NS_SWIFT_NAME(Options)
 /**
  * Array of integrations to install.
  */
-@property (nullable, nonatomic, copy) NSArray<NSString *> *integrations;
+@property (nullable, nonatomic, readonly) NSArray<NSString *> *integrations;
```
Comment (brustolin) [inline Sources/Sentry/Public/SentryOptions.h:167]: h: Why did we changed this in here?
Integrations can be manipulated by users to remove something they don’t like.

File: Sources/Swift/Integrations/UserFeedback/SentryUserFeedbackIntegrationDriver.swift [hunk 2]
```diff
@@ -7,19 +7,22 @@ import UIKit
  * An integration managing a workflow for end users to report feedback via Sentry.
  * - note: The default method to show the feedback form is via a floating widget placed in the bottom trailing corner of the screen. See the configuration classes for alternative options.
  */
-class SentryUserFeedbackIntegration {
+@objcMembers public class SentryUserFeedbackIntegrationDriver: NSObject {
```
Comment (brustolin) [inline Sources/Swift/Integrations/UserFeedback/SentryUserFeedbackIntegrationDriver.swift:10]: h: By making it public, this will be exposed to the user. Is that the intention? If so, how should the user interact with it directly?

File: Sources/Swift/Integrations/UserFeedback/SentryUserFeedbackIntegration.m [hunk 3]
```diff
@@ -0,0 +1,14 @@
+#import "SentryUserFeedbackIntegration.h"
+#import "SentryOptions+Private.h"
+#import "SentrySwift.h"
+
+@implementation SentryUserFeedbackIntegration {
```
Comment (brustolin) [inline Sources/Swift/Integrations/UserFeedback/SentryUserFeedbackIntegration.m:5]: l: Do you plan to add more features to this class?
Why not use `SentryUserFeedbackIntegrationDriver` as the integration like before?

File: Sources/Swift/Integrations/UserFeedback/SentryUserFeedbackWidget.swift [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+            fatalError("init(coder:) has not been implemented")
+        }
+        
+        override func hitTest(_ point: CGPoint, with event: UIEvent?) -> UIView? {
+            guard let result = super.hitTest(point, with: event) else {
+                return nil
+            }
+            guard result.isKind(of: UIButton.self) else {
+                return nil
+            }
```
Comment (brustolin) [inline Sources/Swift/Integrations/UserFeedback/SentryUserFeedbackWidget.swift:86]: l: ```suggestion
            guard let result = super.hitTest(point, with: event), !result.isKind(of: UIButton.self) else {
                return nil
            }
```

**Comment by brustolin** [comment]
I think we should detect TabBar to arrange the button according.

Simulator Screenshot - iPhone 16 Pro - 2024-09-24 at 09 15 13

**Comment by brustolin** [comment]
h: We can debate it later, but currently we're organising Swift code under Swift folder, so

|This|Should be here|
|-|-|
|Screenshot 2024-09-24 at 09 28 55|Screenshot 2024-09-24 at 09 29 06|

File: Sources/Sentry/Public/SentryOptions.h (hunk 1 above)
Comment (armcknight) [reply Sources/Sentry/Public/SentryOptions.h:167]: I wanted to be able to mutate the list in the implementation.

I misunderstood `copy`, thinking it _hands out copies on access_, but it actually _makes copies on assignment_... that's ok, because I should be able to make these readwrite/copy and it should still work for my purposes.

File: Sources/Swift/Integrations/UserFeedback/SentryUserFeedbackIntegration.m (hunk 3 above)
Comment (armcknight) [reply Sources/Swift/Integrations/UserFeedback/SentryUserFeedbackIntegration.m:5]: Originally I had it like that, but it wasn't compiling with a swift integration class, and we don't have any others like that. I essentially tried to copy what you have for replays.

File: Sources/Swift/Integrations/UserFeedback/SentryUserFeedbackWidget.swift (hunk 4 above)
Comment (armcknight) [reply Sources/Swift/Integrations/UserFeedback/SentryUserFeedbackWidget.swift:86]: Yes, I'll have to add a boolean property there to allow hits to the form while it's displayed. That'll come in a separate PR for implementing the form (and I have implementations of things I've done this for in the past that I can reference)

Comment (armcknight) [reply Sources/Swift/Integrations/UserFeedback/SentryUserFeedbackWidget.swift:86]: I usually prefer leaving separate conditions in separate guards instead of using comma-delimited lists, helps debugging and logging later if need be

File: Sources/Swift/Integrations/UserFeedback/SentryUserFeedbackWidget.swift [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+                label.text = "Hi, I'm a user feedback form!"
+                formDialog.view.addSubview(label)
+                label.translatesAutoresizingMaskIntoConstraints = false
+                label.textAlignment = .center
+                NSLayoutConstraint.activate([
+                    label.leadingAnchor.constraint(equalTo: formDialog.view.leadingAnchor),
+                    label.trailingAnchor.constraint(equalTo: formDialog.view.trailingAnchor),
+                    label.centerYAnchor.constraint(equalTo: formDialog.view.centerYAnchor),
+                ])
+                present(formDialog, animated: true)
```
[the remark this answers is not in the document]
Comment (armcknight) [reply Sources/Swift/Integrations/UserFeedback/SentryUserFeedbackWidget.swift:66]: Yep I added the animations idea last minute and haven't completed it yet.

**Comment by armcknight** [comment]
> I think we should detect TabBar to arrange the button according.
> 
> Simulator Screenshot - iPhone 16 Pro - 2024-09-24 at 09 15 13

This should be on the consumer to use the `layoutUIOffset` and later, enabling draggability of the button will obviate this. I don't think trying to predict all the possible combinations of view hierarchy, or spelunking through them, is wise. That being said I'm not sure which sample app or device you're testing, because I thought I already fixed this with the sample apps' usage of said `layoutUIOffset` property 🤔

File: Sources/Sentry/Public/SentryOptions.h (hunk 1 above)
Comment (armcknight) [reply Sources/Sentry/Public/SentryOptions.h:167]: Hm, sadly I cannot make it a private readwrite NSMutableArray and public copy nonmutable NSArray

Comment (armcknight) [reply Sources/Sentry/Public/SentryOptions.h:167]: I found another way to do it

**Comment by brustolin** [comment]
> That being said I'm not sure which sample app or device you're testing, because I thought I already fixed this with the sample apps' usage of said layoutUIOffset property 🤔

I just downloaded this branch.

**Comment by armcknight** [comment]
> > That being said I'm not sure which sample app or device you're testing, because I thought I already fixed this with the sample apps' usage of said layoutUIOffset property 🤔
> 
> I just downloaded this branch.

> which sample app or device

**Comment by brustolin** [comment]
> which sample app or device

iOS-Swift sample
iPhone 16 Pro

**Comment by armcknight** [comment]
The diff is currently a little over 500 lines added, note that half of that is just the UIBezier drawing the megaphone that I got from our SVG design asset :D

File: Sources/Swift/Integrations/UserFeedback/SentryUserFeedbackIntegrationShell.h [hunk 6]
```diff
@@ -0,0 +1,9 @@
+#import "SentryBaseIntegration.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface SentryUserFeedbackIntegrationShell : SentryBaseIntegration
```
Comment (armcknight) [inline Sources/Swift/Integrations/UserFeedback/SentryUserFeedbackIntegrationShell.h:5]: So, in the previous PR, I made `SentryUserFeedbackIntegration` a swift file, before I knew that it isn't currently possible to implement integrations in Swift (AFAICT anyway). But, renaming that file/class so that this objc class could have that name made the diff of that file messy, so for now I'm making this a placeholder name, and I'll do the renames in a separate PR.

**Comment by brustolin** [comment]
For default values, we should respect iOS assessibility values:

Simulator Screenshot - iPhone 16 Pro - 2024-10-16 at 08 42 45

Simulator Screenshot - iPhone 16 Pro - 2024-10-16 at 08 44 10

This:
image 

Should be:
image

In this configuration

**Review by brustolin — CHANGES_REQUESTED** [review]
[This](https://github.com/getsentry/sentry-cocoa/pull/4364#issuecomment-2370418184) need to be addressed before merging the PR.

**Review by brustolin — COMMENTED** [review]
I didn't review the code yet, just how to use it and the button itself. 

A couple of extra things:
- Need right-to-left support
- I think configureTheme, should be inside configureWidget, or at least the properties that configure the widget, it took me an extra second to find this out, when it would be much easier to set the widget theme inside the configureWidget callback

File: Sources/Swift/Integrations/UserFeedback/SentryUserFeedbackWidgetButtonMegaphoneIconView.swift [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+            if UIScreen.main.traitCollection.userInterfaceStyle == .dark {
+                svgLayer.fillColor = config.darkTheme.foreground.cgColor
+            } else {
+                configureLightTheme()
+            }
+        } else {
+            configureLightTheme()
+        }
+        
+        svgLayer.fillRule = .evenOdd
```
Comment (brustolin) [inline Sources/Swift/Integrations/UserFeedback/SentryUserFeedbackWidgetButtonMegaphoneIconView.swift:23]: l: This is just a suggestion to use alongside the megaphoneShape suggestion.
Feel free to ignore it 

```suggestion
        svgLayer.path = megaphoneShape
        svgLayer.lineWidth = 1.5
        svgLayer.lineJoin = .round
        svgLayer.fillColor = UIColor.clear.cgColor
        
        func configureLightTheme() {
            svgLayer.strokeColor = config.theme.foreground.cgColor
        }
        if #available(iOS 12.0, *) {
            if UIScreen.main.traitCollection.userInterfaceStyle == .dark {
                svgLayer.fillColor = config.darkTheme.foreground.cgColor
            } else {
                configureLightTheme()
            }
        } else {
            configureLightTheme()
        }
        
```

File: Sources/Swift/Integrations/UserFeedback/SentryUserFeedbackWidgetButtonMegaphoneIconView.swift [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+            layer.transform = CATransform3DMakeScale(config.scaleFactor, config.scaleFactor, 0)
+        }
+    }
+    
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    //swiftlint:disable function_body_length
+    lazy var megaphoneShape: CGPath = {
```
Comment (brustolin) [inline Sources/Swift/Integrations/UserFeedback/SentryUserFeedbackWidgetButtonMegaphoneIconView.swift:38]: l: I love the idea of drawing like this.

You can achieve a very similar result with the following code:

```Swift
lazy var megaphoneShape: CGPath = {
        let path = CGMutablePath()
        
        path.move(to: CGPoint(x: 1, y: 3))
        path.addLine(to: CGPoint(x: 7, y: 3))
        path.addLine(to: CGPoint(x: 10, y: 1))
        path.addLine(to: CGPoint(x: 12, y: 1))
        path.addLine(to: CGPoint(x: 12, y: 11))
        path.addLine(to: CGPoint(x: 10, y: 11))
        path.addLine(to: CGPoint(x: 7, y: 9))
        path.addLine(to: CGPoint(x: 1, y: 9))
        path.closeSubpath()
        
        path.addRect(CGRect(x: 2, y: 9, width: 3.5, height: 6))
        
        path.move(to: CGPoint(x: 12, y: 6))
        path.addRelativeArc(center: CGPoint(x: 12, y: 6), radius: 3, startAngle: -(.pi/2), delta: .pi)
        
        return path
    }
```

**Review by brustolin — COMMENTED** [review]
Just a suggestion for the megaphone drawing

**Comment by armcknight** [comment]
@brustolin accessibility sizes and right-to-left languages are known tasks: https://github.com/getsentry/sentry-cocoa/issues/4270#issuecomment-2408210459

I already have right-to-left support going after i linked the PR last night, see the latest update in the issue: https://github.com/getsentry/sentry-cocoa/issues/4270#issuecomment-2417077633 with the implementation work in https://github.com/getsentry/sentry-cocoa/pull/4364/commits/abc51191c823c86b6cdf09317f14746e728b3e9d

**Comment by armcknight** [comment]
> I think configureTheme, should be inside configureWidget, or at least the properties that configure the widget, it took me an extra second to find this out, when it would be much easier to set the widget theme inside the configureWidget callback

This can be debated separately from this PR if people want to. I'll just note here that theme options will also be needed in the form, so I tried creating a composition hierarchy that indicates they are used by more than just the widget (button):
Screenshot 2024-10-16 at 11 07 56 AM

Since the widget and form will use a lot of the same options I don't think it makes sense to make separate theme objects for each to own separately.

File: Sources/Swift/Integrations/UserFeedback/SentryUserFeedbackWidgetButtonMegaphoneIconView.swift (hunk 8 above)
Comment (armcknight) [reply Sources/Swift/Integrations/UserFeedback/SentryUserFeedbackWidgetButtonMegaphoneIconView.swift:45]: I will let @Jesse-Box or @vuluongj20 comment on this.

**Comment by armcknight** [comment]
> [This](https://github.com/getsentry/sentry-cocoa/pull/4364#issuecomment-2370418184) need to be addressed before merging the PR.

@brustolin Would you mind if we do it in a separate PR along with https://github.com/getsentry/sentry-cocoa/pull/4364#discussion_r1787018390? I just don't want to mess the diff up here. Nothing is being shipped yet so there's no danger of a file not being delivered where it should be.

File: Sources/Swift/Integrations/UserFeedback/SentryUserFeedbackWidgetButtonMegaphoneIconView.swift (hunk 8 above)
Comment (armcknight) [reply Sources/Swift/Integrations/UserFeedback/SentryUserFeedbackWidgetButtonMegaphoneIconView.swift:48]: I think i'm just gonna go with the simpler version and if anyone doesn't like it, we can always revert. Thanks @brustolin

**Comment by brustolin** [comment]
@brustolin yes, then plan is to merge to main for each PR. The option is currently in SentryOptions+Private so they should not be able to get a widget showing in their app.

Ok, good enough for me. Thanks!

File: Sources/Sentry/SentryOptions.m [hunk 9]
```diff
@@ -39,35 +39,33 @@ @implementation SentryOptions {
     BOOL _enableTracingManual;
 }
 
-+ (NSArray<NSString *> *)defaultIntegrations
+NSMutableArray<NSString *> *
+sentry_defaultIntegrations(void)
```
Comment (brustolin) [inline Sources/Sentry/SentryOptions.m:43]: l: Why was this changed?
Just curious.

**Review by brustolin — APPROVED** [review]
I have my reservations about the composition of theme configuration, but this is not a blocker. 

LGTM

File: Sources/Sentry/SentryOptions.m (hunk 9 above)
Comment (armcknight) [reply Sources/Sentry/SentryOptions.m:43]: I changed it to return a mutable array so that I can conditionally add the user feedback integration before installing all integrations, if the options property to configure its options is set (which implicitly enables the feature)
